android/src/com/mapswithme/maps/location/SocketWrapper.java

```diff
@@ -0,0 +1,161 @@
+package com.mapswithme.maps.location;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.mapswithme.util.log.DebugLogger;
+import com.mapswithme.util.log.Logger;
+
+import javax.net.SocketFactory;
+import javax.net.ssl.SSLSocketFactory;
+import java.io.IOException;
+import java.net.Socket;
+import java.net.SocketException;
+
+/**
+ * Implements {@link PlatformSocket} interface that will be used by the core for
+ * sending/receiving the raw data trough platform socket interface.
+ * <p>
+ * The instance of this class is supposed to be created in JNI layer
+ * and supposed to be used in the thread safe environment, i.e. thread safety
+ * should be provided externally (by the client of this class).
+ * <p>
+ * <b>All public methods are blocking and shouldn't be called from the main thread.</b>
+ */
+class SocketWrapper implements PlatformSocket
+{
+  private final static Logger sLogger = new DebugLogger(SocketWrapper.class.getSimpleName());
+  @Nullable
+  private Socket mSocket;
+  @Nullable
+  private String mHost;
+  private int mPort;
+
+  @Override
+  public boolean open(@NonNull String host, int port)
+  {
+    if (mSocket != null)
+    {
+      sLogger.e("Socket is already opened. Seems that it wasn't be closed.");
+      return false;
+    }
+
+    mHost = host;
+    mPort = port;
+
+    Socket socket = createSocket(host, port, true);
+    if (socket != null && socket.isConnected())
+      mSocket = socket;
+
+    return mSocket != null;
+  }
+
+  @Nullable
+  private static Socket createSocket(@NonNull String host, int port, boolean ssl)
+  {
+    if (ssl)
+    {
+      SocketFactory sf = SSLSocketFactory.getDefault();
+      try
+      {
+        return sf.createSocket(host, port);
+      } catch (IOException e)
+      {
+        sLogger.e("Failed to create the ssl socket, mHost", host, " mPort = ", port, e);
+      }
+    } else
+    {
+      try
+      {
+        return new Socket(host, port);
+      } catch (IOException e)
+      {
+        sLogger.e("Failed to create the socket, mHost = ", host, " mPort = ", port, e);
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public boolean close()
+  {
+    if (mSocket == null)
+    {
+      sLogger.e("Socket is already closed or it wasn't be opened before");
+      return false;
+    }
+
+    try
+    {
+      mSocket.close();
+      return true;
+    } catch (IOException e)
+    {
+      sLogger.e("Failed to close socket: ", this, e);
+    } finally
+    {
+      mSocket = null;
```

Comment:
So, according implementation of method Socket.close() the socket will be marked as closed regardless whether an IOexception was thrown or not. That's why, we should nullify this field (Socket), because we have a contract with the core design, i.e. after each 'close' the Java side will wait for 'open' call. Since there is nullcheck at the start of 'open' method we should nullify the socket in 'close' method.

If it's not clear, please let me know